kernel/cClosure.ml

```diff
@@ -595,43 +614,52 @@ let rec to_constr lfts v =
         mkProj (p,to_constr lfts c)
 
     | FLambda (len, tys, f, e) ->
-      if is_subs_id e && is_lift_id lfts then
+      let is_id1 = is_subs_id (fst e) && is_lift_id lfts in
+      if is_id1 && Instance.is_empty (snd e) then
         Term.compose_lam (List.rev tys) f
+      else if is_id1 then
+        subst_instance_constr (snd e) (Term.compose_lam (List.rev tys) f)
       else
         let subs = comp_subs lfts e in
-        let tys = List.mapi (fun i (na, c) -> na, subst_constr (subs_liftn i subs) c) tys in
-        let f = subst_constr (subs_liftn len subs) f in
+        let tys = List.mapi (fun i (na, c) -> na, subst_constr (subs_liftn i subs, snd e) c) tys in
+        let f = subst_constr (subs_liftn len subs, snd e) f in
         Term.compose_lam (List.rev tys) f
     | FProd (n,t,c)   ->
         mkProd (n, to_constr lfts t,
                    to_constr (el_lift lfts) c)
     | FLetIn (n,b,t,f,e) ->
-      let subs = comp_subs (el_lift lfts) (subs_lift e) in
+      let subs = comp_subs (el_lift lfts) (fstapp subs_lift e) in
         mkLetIn (n, to_constr lfts b,
                     to_constr lfts t,
-                    subst_constr subs f)
+                    subst_constr (subs, snd e) f)
     | FEvar ((ev,args),env) ->
       let subs = comp_subs lfts env in
-        mkEvar(ev,Array.map (fun a -> subst_constr subs a) args)
+        mkEvar(ev,Array.map (fun a -> subst_constr (subs, snd env) a) args)
     | FLIFT (k,a) -> to_constr (el_shft k lfts) a
     | FCLOS (t,env) ->
-      if is_subs_id env && is_lift_id lfts then t
+      let is_id1 = is_subs_id (fst env) && is_lift_id lfts in
+      if is_id1 && Instance.is_empty (snd env) then
+ t
+      else if is_id1 then
+ subst_instance_constr (snd env) (t)
       else
         let subs = comp_subs lfts env in
-        subst_constr subs t
+        subst_constr (subs, snd env) t
     | FLOCKED -> assert false (*mkVar(Id.of_string"_LOCK_")*)
 
-and subst_constr subst c = match Constr.kind c with
+and subst_constr (subst, inst) c = match Constr.kind c with
```

Comment:
I think you should curry this pair away.

Reply:
it's used as an argument to some combinators (`map_with_binders, Array.Fun1.map`) so I would probably have to uncurry the function in these places, wouldn't I?